relay-rust/src/cli_args.rs

```diff
@@ -14,168 +14,256 @@
+            SubCommand::with_name("tunnel")
+                .about("Set up the 'adb reverse' tunnel")
+                .long_about("Set up the 'adb reverse' tunnel.{n}    \
+                Note: If a device is unplugged then plugged back while gnirehtet is{n}\
+                active, resetting the tunnel is sufficient to get the{n}\
+                connection back.",
+                )
+                .arg(&serial)
+                .arg(&port),
+        )
+        .subcommand(
+            SubCommand::with_name("relay")
+                .about("Start the relay server in the current terminal.")
+                .arg(&port),
+        )
+}
 
-    #[test]
-    fn test_serial_and_dns_servers() {
-        let raw_args = vec!["myserial", "-d", "8.8.8.8"];
-        let args = CommandLineArguments::parse(ACCEPT_ALL, raw_args).unwrap();
-        assert_eq!("myserial", args.serial.unwrap());
-        assert_eq!("8.8.8.8", args.dns_servers.unwrap());
-    }
+#[derive(Clone, Default)]
+pub(crate) struct Args {
+    pub(crate) serial: Option<String>,
+    pub(crate) dns_servers: Option<String>,
+    pub(crate) routes: Option<String>,
+    pub(crate) port: u16,
+}
 
-    #[test]
-    fn test_dns_servers_and_serial() {
-        let raw_args = vec!["-d", "8.8.8.8", "myserial"];
-        let args = CommandLineArguments::parse(ACCEPT_ALL, raw_args).unwrap();
-        assert_eq!("myserial", args.serial.unwrap());
-        assert_eq!("8.8.8.8", args.dns_servers.unwrap());
+impl Args {
+    pub(crate) fn serial(&self) -> Option<&str> {
+        self.serial.as_deref()
     }
-
-    #[test]
-    fn test_serial_with_no_dns_servers_parameter() {
-        let raw_args = vec!["myserial", "-d"];
-        assert!(CommandLineArguments::parse(ACCEPT_ALL, raw_args).is_err());
+    pub(crate) fn routes(&self) -> Option<&str> {
+        self.routes.as_deref()
     }
-
-    #[test]
-    fn test_no_dns_servers_parameter() {
-        let raw_args = vec!["-d"];
-        assert!(CommandLineArguments::parse(ACCEPT_ALL, raw_args).is_err());
+    pub(crate) fn dns_servers(&self) -> Option<&str> {
+        self.dns_servers.as_deref()
     }
-
-    #[test]
-    fn test_routes_parameter() {
-        let raw_args = vec!["-r", "1.2.3.0/24"];
-        let args = CommandLineArguments::parse(ACCEPT_ALL, raw_args).unwrap();
-        assert_eq!("1.2.3.0/24", args.routes.unwrap());
+    pub(crate) fn port(&self) -> u16 {
+        self.port
     }
+}
 
-    #[test]
-    fn test_no_routes_parameter() {
-        let raw_args = vec!["-r"];
-        assert!(CommandLineArguments::parse(ACCEPT_ALL, raw_args).is_err());
+impl<'a, 'b> From<&'a ArgMatches<'b>> for Args {
+    fn from(m: &'a ArgMatches<'b>) -> Self {
+        Args {
+            serial: m.value_of("serial").map(ToOwned::to_owned),
+            dns_servers: m.value_of("dns-servers").map(ToOwned::to_owned),
+            routes: m.value_of("routes").map(ToOwned::to_owned),
+            port: value_t_or_exit!(m.value_of("port"), u16),
```

Comment:
I suppose this should not be written in the current way, as some sub-commands does not even have `port` defined, even though `port` has a default value?